src/compiler/checker.ts

```diff
@@ -7336,7 +7395,12 @@ namespace ts {
             if (needToCaptureLexicalThis) {
                 captureLexicalThis(node, container);
             }
-
+            if (isFunctionLike(container)) {
+                const signature = getSignatureFromDeclaration(container);

```

Comment:
This isn't going to work when the containing function has a contextually typed `this` because the signature will appear to have no `this` type. I think you need to `getTypeOfSymbol(getSymbolOfNode(container))` and then first the `this` type out of the signatures of the type.